.github/scripts/coverage_report_breakdown.py

```diff
@@ -0,0 +1,174 @@
+                                       for d in coverage_data)
+        metrics_dict["percent_covered"] = metrics_dict[
+            "covered_lines"] / metrics_dict["num_statements"] * 100
+        return CoverageMetrics(metrics_dict)
+
+    def tests_metrics(self) -> CoverageMetrics:
+        """
+        creates summary metrics for all tests
+        """
+        return self._calculate_metrics(self.tests.values)
+
+    def source_metrics(self, sub_dir: Optional[str] = None) -> CoverageMetrics:
+        """
+        creates summary metrics for the requested vllm subdirectory,
+        or for the reported vllm source if a subdirectory is not specified.
+        sub_dir = "" will report for files directly under vllm
+        """
+        data = self.source
+        if sub_dir is not None:
+            data = self.source[self.source["sub_dir"] == sub_dir]
+
+        return self._calculate_metrics(data.values)
+
+    def to_github_markdown(self) -> str:
+        """
+        returns a string in the form of github compatible markdown with top
+        level and drill down metrics.
+        """
+        # make a dataframe with top level metric summary info
+        overall_metrics = self.overall_metrics
+        overall_metrics["Collection"] = "Overall"
+        test_metrics = self.tests_metrics()
+        test_metrics["Collection"] = "Test Code"
+        source_metrics = self.source_metrics()
+        source_metrics["Collection"] = "Source Code"
+        summary_df = pd.DataFrame(
+            [overall_metrics, test_metrics, source_metrics])
+        # make percent_covered value compatible with the string "%" formatting
+        summary_df["percent_covered"] = summary_df["percent_covered"] / 100
+
+        # compose a set of the subdirectory breakdown summary info
+        breakdown_list = []
+        for sub_dir in sorted(cc.source["sub_dir"].unique()):
+            sub_dir_metrics = cc.source_metrics(sub_dir)
+            label = "vllm 'root'" if sub_dir == "" else sub_dir
+            sub_dir_metrics["Collection"] = label
+            breakdown_list.append(sub_dir_metrics)
+        breakdown_df = pd.DataFrame(breakdown_list)
+        # make percent_covered value compatible with the string "%" formatting
+        breakdown_df["percent_covered"] = breakdown_df["percent_covered"] / 100
+
+        # join the top level and breakdown data with separator rows between them
+        # add a separator row and subtitle row
+        empty_row_df = pd.Series(
+            pd.NA, index=summary_df.columns).to_frame().transpose()
+        header_row_df = empty_row_df.copy()
+        header_row_df["Collection"] = "vllm Subdirs"
+        summary_df = pd.concat(
+            [summary_df, empty_row_df, header_row_df, breakdown_df],
+            ignore_index=True)
+        # clean up the `nan` values for display purposes
+        summary_df = summary_df.astype(str)
+        summary_df.replace({"nan": None}, inplace=True)
```

Comment:
quick question, how do we generate "nan"?

Reply:
this was "fun" 😉 .  to get that empty row and header row between the overall info and the per-sub-directory breakdown I needed to add that `empty_row_df` and `header_row_df`.  To get the `concat` to work successfully I needed to populate those with `pd.NA`, (which is `nan`).  once the `concat` was done, I could clean up those `nan`s